draft-ietf-taps-impl.md

```diff
@@ -146,10 +146,13 @@ The connection objects that are exposed to applications for Transport Services a
 - the Connection, the basic object that represents a flow of data as Messages in either direction between the Local and Remote Endpoints;
 - and the Listener, a passive waiting object that delivers new Connections.
 
-Preconnection objects should be implemented as bundles of properties that an application can both read and write. Once a Preconnection has been used to create an outbound Connection or a Listener, the implementation should ensure that the copy of the properties held by the Connection or Listener is immutable. This may involve performing a deep-copy, copying the object with all the objects it references, if the application is still able to modify properties on the original Preconnection object.
+Preconnection objects should be implemented as bundles of properties that an application can both read and write. An implementation could allow properties of a preconnection to be updated, but changes to the Preconnection Selection and Connection may be language dependent and is to be avoided in code that could be used across different implementations.
 
```

Comment:
The original seems clearer to me: it's more explicit about what happens, and when.

Reply:
Oh dear we don't seem to agree - I think the original hints at everything being immutable. Instead, I think everything should be copied; but only the selection params are immutable, the other params are mutable on the copy.

Reply:
That would be fine, but I didn't get that meaning from the new text.

Reply:
So... what's missing? - The text on copying was moved to the section after "once a connection is established.

Reply:
I can't parse "An implementation could allow properties of a preconnection to be updated, but changes to the Preconnection Selection and Connection may be language dependent and is to be avoided in code that could be used across different implementations". I read it as trying to simplify the original wording, but actually making it less clear to me. I don't see how it relates to the selection parameters being immutable and while the other parameters remain mutable.

Reply:
Improved text

Reply:
The last two sentences here still seem vague and contradictory with the below. I think it'd be clearer just to remove them, and leave the detailed following explanations.

Reply:
@martinduke :  The expected operation is now clearer, csp suggests simply not mentioning the "update during pre-connection" possibility. What do you think?

Reply:
I have to agree with @csperkins that the original was clearer, mainly for being more explicit - but I think that, after looking at this for some time, I now see where the problem is (which @csperkins is perhaps missing, and which @gorryfair is perhaps trying to fix).

I suspect that @gorryfair wants to avoid that we say, here, that properties on a Connection cannot be changed after creating an outbound Connection or a Listener. This would, of course, be wrong, but the text says that they are all "immutable", which gives a wrong indication.

Then again, the proposed text now doesn't look like a good fix to me. E.g., it says "an implementation could allow applications to...", followed by "however, applications should not do this". This is awkward. Maybe you meant to write "Connection", not "preconnection" in the  second sentence?

Here's a proposal: a change to the original text that avoids falsely indicating that everything becomes immutable:

***
Preconnection objects should be implemented as bundles of properties that an application can both read and write. A Preconnection object influences a Connection only at one point in time: when the Connection is created. Once a Preconnection has been used to create an outbound Connection or a Listener, the implementation should ensure that the copy of the properties held by the Connection or Listener is not affected when the application makes changes to the Preconnection object. This may involve performing a deep-copy, copying the object with all the objects it references.
***

I also removed the trailing "if the application is still able..." because I don't think this added clarity.

Reply:
I tried, and merged. Please check the new diff and suggest improvements.

Reply:
FWIW, I agree with this update. Thanks!